icicle_v3/include/icicle/fields/host_math.h

```diff
@@ -112,18 +174,52 @@ namespace host_math {
     }
   }
 
-  template <unsigned NLIMBS, bool SUBTRACT, bool CARRY_OUT>
-  static constexpr HOST_INLINE uint32_t
-  add_sub_limbs(const storage<NLIMBS>& xs, const storage<NLIMBS>& ys, storage<NLIMBS>& rs)
+  template <unsigned NLIMBS_A, unsigned NLIMBS_B = NLIMBS_A>
+  static HOST_INLINE void multiply_raw_64(const uint64_t* a, const uint64_t* b, uint64_t* r)
   {
-    const uint32_t* x = xs.limbs;
-    const uint32_t* y = ys.limbs;
-    uint32_t* r = rs.limbs;
-    uint32_t carry = 0;
-    carry_chain<NLIMBS, false, CARRY_OUT> chain;
-    for (unsigned i = 0; i < NLIMBS; i++)
-      r[i] = SUBTRACT ? chain.sub(x[i], y[i], carry) : chain.add(x[i], y[i], carry);
-    return CARRY_OUT ? carry : 0;
+    for (unsigned i = 0; i < NLIMBS_B / 2; i++) {
+      uint64_t carry = 0;
+      for (unsigned j = 0; j < NLIMBS_A / 2; j++)
+        r[j + i] = host_math::madc_cc_64(a[j], b[i], r[j + i], carry);
+      r[NLIMBS_A / 2 + i] = carry;
+    }
+  }
+
+  template <unsigned NLIMBS_A, unsigned NLIMBS_B = NLIMBS_A>
+  static HOST_INLINE void
+  multiply_raw_64(const storage<NLIMBS_A>& as, const storage<NLIMBS_B>& bs, storage<NLIMBS_A + NLIMBS_B>& rs)
+  {
+    const uint64_t* a = as.limbs64;
+    const uint64_t* b = bs.limbs64;
+    uint64_t* r = rs.limbs64;
+    multiply_raw_64<NLIMBS_A, NLIMBS_B>(a, b, r);
+  }
+
+  template <unsigned NLIMBS_A, unsigned NLIMBS_B = NLIMBS_A, bool USE_32 = false>
+  static constexpr HOST_INLINE void
+  multiply_raw(const storage<NLIMBS_A>& as, const storage<NLIMBS_B>& bs, storage<NLIMBS_A + NLIMBS_B>& rs)
+  {
+    static_assert(
+      (NLIMBS_A % 2 == 0 || NLIMBS_A == 1) && (NLIMBS_B % 2 == 0 || NLIMBS_B == 1),
+      "odd number of limbs is not supported\n");
+    if constexpr (USE_32) {
+      multiply_raw_32<NLIMBS_A, NLIMBS_B>(as, bs, rs);
+      return;
+    } else if constexpr ((NLIMBS_A == 1 && NLIMBS_B == 2) || (NLIMBS_A == 2 && NLIMBS_B == 1)) {
+      multiply_raw_32<NLIMBS_A, NLIMBS_B>(as, bs, rs);
+      return;
+    } else if constexpr (NLIMBS_A == 1 && NLIMBS_B == 1) {
+      rs.limbs[0] = host_math::madc_cc(as.limbs[0], bs.limbs[0], 0, rs.limbs[1]);
```

Comment:
(1) this madc_cc is doing a redundant +0. right?
(2) if rs.limbs[1] is not zero, it is contributing to the result as carry in. Is that a bug?

Same for madc_cc_64

Reply:
fixed